### PR TITLE
Unsafe.Unlikely (Assume) draft implementation

### PR DESCRIPTION
**Pull request by EgorBo**
Just a quick draft for https://github.com/dotnet/runtime/issues/4966 for both RyuJIT CoreCLR and Mono-LLVM (maybe it will help to push it forward since it's 4 years old).

It allows programmers to give hints to JIT about branches' probabilities, e.g. to avoid goto hacks [like this](https://github.com/dotnet/runtime/blob/469ece74b6e5b7e94991dde21b3c2409194ceab9/src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs#L523) in perf critical code.

#### Sample:
```csharp
static int Foo(int a)
{
    if (Unsafe.Unlikely(a <= 0))   // hint: `a` is unlikely to be <= 0
        return a * a; 

    return 0;
}
```
#### Codegen without `Unlikely`:
```asm
       85D2                 test     edx, edx
       7F06                 jg       SHORT G_M36845_IG05
       8BC2                 mov      eax, edx
       0FAFC2               imul     eax, edx
       C3                   ret                    ;; return a * a
G_M36845_IG05:
       8BC2                 mov      eax, edx
       C3                   ret                    ;; return 0
```
#### Codgen with `Unlikely` (this is what this PR emits)
```asm
       85D2                 test     edx, edx
       7E03                 jle      SHORT G_M4270_IG05
       8BC2                 mov      eax, edx
       C3                   ret                     ;; return 0
G_M4270_IG05:
       8BC2                 mov      eax, edx
       0FAFC2               imul     eax, edx 
       C3                   ret                     ;; return a * a 
```

Also, I implemented it for Mono-LLVM (I use `@llvm.expect.i32` intrinsic, which is emitted when you use `__builtin_expect` in C/C++)
#### Mono-LLVM without `Unlikely` (macOS)
```asm
	mov	ecx, edi
	imul	ecx, ecx   ;; Mono-LLVM emits branchless code for our sample
	xor	eax, eax   
	test	edi, edi
	cmovle	eax, ecx   
	ret
```
#### Mono-LLVM with `Unlikely` (macOS)
```asm
	xor	eax, eax
	test	edi, edi
	jle	1 <gram_Foo1__int_+0x7>   ;; branch is better than cmove in this case
	ret                               ;; return 0
	imul	edi, edi  
	mov	eax, edi
	ret                               ;; return a * a
```
cc @jkotas

### CODE REVIEW

File: src/mono/mono/mini/intrinsics.c [hunk 1]
```diff
@@ -588,6 +588,17 @@ emit_unsafe_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatu
 		MONO_ADD_INS (cfg->cbb, ins);
 		return ins;
 	}
+	else if (!strcmp (cmethod->name, "Unlikely") && COMPILE_LLVM (cfg)) {
+		g_assert (fsig->param_count == 1);
```
Comment (vargaz) [inline src/mono/mono/mini/intrinsics.c]: Indentation.
COMPILE_LLVM () should come first.

Comment (vargaz) [inline src/mono/mono/mini/intrinsics.c]: Indentation.
COMPILE_LLVM (cfg) should come first.

**Comment by vargaz** [comment]
The mono parts look ok to me.
Shouldn't there be a Likely() as well ?

**Comment by EgorBo** [comment]
> The mono parts look ok to me.
> Shouldn't there be a Likely() as well ?

I guess it's more an API design question, e.g. in the original issue it's proposed to be `Assume(x, y)`. So it needs API review first.

**Comment by stephentoub** [comment]
> Just a quick draft for #4966 for both RyuJIT CoreCLR and Mono-LLVM

Thanks.  Is there any reason this need to be a PR?  Can you not just link to the relevant commit from that issue?  This doesn't seem actionable.

**Comment by EgorBo** [comment]
> > Just a quick draft for #4966 for both RyuJIT CoreCLR and Mono-LLVM
> 
> Thanks. Is there any reason this need to be a PR? Can you not just link to the relevant commit from that issue? This doesn't seem actionable.

Ok, will close it if there won't be any interest in some short term. Next time will use commits instead if you don't mind.

**Comment by stephentoub** [comment]
> Ok, will close it if there won't be any interest in some short term. Next time will use commits instead if you don't mind.

Thanks.  Mainly I want to keep the PR list to things that folks should act on by getting them merged or closed.  We're currently pushing 190 PRs, and from I can see, most folks are ignoring anything after the first page or two.  I'd love to drive down the number, and getting there means not having unactionable PRs sitting around purely for discussion purposes.  For non-committed ideas, I'd like for those discussions to be had on the relevant issues; linking to commits from those issues is a good thing to do, of course.

**Comment by CoffeeFlux** [comment]
@stephentoub if you to close an issue in dotnet/runtime dealing with Mono code, please try to close the corresponding mono/mono mirrored PR as well to keep things tidy.

**Comment by stephentoub** [comment]
> if you to close an issue in dotnet/runtime dealing with Mono code, please try to close the corresponding mono/mono mirrored PR as well to keep things tidy.

Ok, will do. If the mirror is able to open PRs, it'd be great if it can close them, too.

**Pull request closed without merging**